File: app/src/main/java/com/naosim/rpgsample/MessageViewModelImpl.java
package com.naosim.rpgsample;

import android.os.Handler;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.TextView;

import com.naosim.rpgmodel.lib.value.Message;
import com.naosim.rpgmodel.lib.viewmodel.MessageViewModel;

import org.jetbrains.annotations.NotNull;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;

/**
 * Created by nao_pillows on 2016/08/06.
 */
public class MessageViewModelImpl implements MessageViewModel {
    private final TextView textView;
    private final Handler handler = new Handler();
    private final View nextIcon;

    public MessageViewModelImpl(TextView textView, View nextIcon) {
        this.textView = textView;
        this.nextIcon = nextIcon;
        dismiss();
    }

    public void setMessageText(final String text) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                textView.setText(text);
            }
        });
    }

    @Override
    public void clear() {
        textView.setText("");
    }

    @Override
    public void appendText(@NotNull Message message, final @NotNull Function0<Unit> onEnd) {
        String text = message.getValue().trim() + "\n";
        final int firstTextCount = textView.getText().toString().length();
        final char data[] = text.toCharArray();

        nextIcon.setVisibility(View.INVISIBLE);
        nextIcon.clearAnimation();

        new Thread(new Runnable() {
            @Override
            public void run() {
                while(true) {
                    String currentText = textView.getText().toString();
                    int i = currentText.length();
                    if(i - firstTextCount >= data.length) {
                        handler.post(new Runnable() {
                            @Override
                            public void run() {
                                nextIcon.setVisibility(View.VISIBLE);
                                AlphaAnimation aanim1 = new AlphaAnimation(1, 0.2f);
                                aanim1.setDuration( 400 );
                                aanim1.setRepeatCount( Animation.INFINITE );
                                aanim1.setRepeatMode( Animation.REVERSE );
                                (nextIcon).startAnimation(aanim1);

                                onEnd.invoke();
                            }
                        });
                        return;
                    }
                    setMessageText(currentText + data[i++ - firstTextCount]);
                    try {
                        Thread.sleep(50L);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                        return;
                    }
                }
            }
        }).start();
    }

    @Override
    public void show() {
        textView.setVisibility(View.VISIBLE);
    }

    @Override
    public void dismiss() {
        textView.setVisibility(View.INVISIBLE);
    }

    @Override
    public boolean isShowing() {
        return textView.getVisibility() == View.VISIBLE;
    }
}

File: app/src/main/kotlin/com/naosim/rpgmodel/lib/android/MessageViewModelImpl.kt
package com.naosim.rpgmodel.lib.android

import android.os.Handler
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.widget.TextView
import com.naosim.rpgmodel.lib.model.value.Message
import com.naosim.rpgmodel.lib.model.viewmodel.MessageViewModel

class MessageViewModelImpl(
        private val textView: TextView,
        private val nextIcon: View
): MessageViewModel {
    private val handler = Handler()

    fun setMessageText(text: String) {
        handler.post { textView.text = text }
    }

    override fun clear() {
        textView.text = ""
    }

    override fun appendText(message: Message, onEnd: () -> Unit) {
        val text = message.value.trim { it <= ' ' } + "\n"
        val firstTextCount = textView.text.toString().length
        val data = text.toCharArray()
        nextIcon.visibility = View.INVISIBLE
        nextIcon.clearAnimation()

        Thread({
            while (true) {
                val currentText = textView.text.toString()
                var i = currentText.length
                if (i - firstTextCount >= data.size) {
                    handler.post {
                        nextIcon.visibility = View.VISIBLE
                        val aanim1 = AlphaAnimation(1f, 0.2f)
                        aanim1.duration = 400
                        aanim1.repeatCount = Animation.INFINITE
                        aanim1.repeatMode = Animation.REVERSE
                        nextIcon.startAnimation(aanim1)
                        onEnd.invoke()
                    }
                    break
                }
                setMessageText(currentText + data[i++ - firstTextCount])
                try {
                    Thread.sleep(50L)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

            }
        }).start()
    }

    override fun show() {
        textView.visibility = View.VISIBLE
    }

    override fun dismiss() {
        textView.visibility = View.INVISIBLE
        nextIcon.clearAnimation()
        nextIcon.visibility = View.INVISIBLE
    }

    override fun isShowing(): Boolean {
        return textView.visibility == View.VISIBLE
    }
}